autodiff/forward/eigen.hpp

```diff
@@ -94,21 +124,116 @@ EIGEN_MAKE_TYPEDEFS_ALL_SIZES(autodiff::dual, dual)
 
 namespace autodiff::forward {
 
+namespace detail {
+/// Compile time foreach for tuples
+template <typename Tuple, typename Callable>
+void forEach(Tuple&& tuple, Callable&& callable)
+{
+    std::apply(
+        [&callable](auto&&... args) { (callable(std::forward<decltype(args)>(args)), ...); },
+        std::forward<Tuple>(tuple)
+    );
+}
+
+/// Create index sequence at interval [Start, End)
+template <std::size_t Start, std::size_t End, std::size_t... Is>
+auto makeIndexSequenceImpl() {
+    if constexpr (End == Start) return std::index_sequence<Is...>();
+    else return makeIndexSequenceImpl<Start, End - 1, End - 1, Is...>();
+}
+
+/// Create tuple view using index sequence
+template <class Tuple, size_t... Is>
+constexpr auto viewImpl(Tuple&& t,
+    std::index_sequence<Is...>) {
+    return std::forward_as_tuple(std::get<Is>(t)...);
+}
+
+/// Create index sequence on [Start, End)
+template <std::size_t Start, std::size_t End>
+using makeIndexSequence = std::decay_t<decltype(makeIndexSequenceImpl<Start, End>())>;
+
+/// Create view on tuple tail size N
+template <size_t N, class Tuple>
+constexpr auto tailView(Tuple&& t) {
+    constexpr auto size = std::tuple_size<std::remove_reference_t<Tuple>>::value;
+    static_assert(N <= size, "N must be smaller or equal than size of tuple");
+    return viewImpl(t, makeIndexSequence<size - N, size>{});
+}
+
+/// Return count of 'joined' tuple elements
+template<typename Tuple>
+auto count(Tuple&& t) -> std::size_t
+{
+    std::size_t n = 0;
+    forEach(t, [&n](auto&& element) {
+        n += element.size();
+    });
+    return n;
+}
+
+/// Wrap T to compatible array interface
+template<typename T>
+struct EigenVectorAdaptor {
+    /// implicit construct from value
+    constexpr EigenVectorAdaptor(T val) : val(val) { }
+    /// operator [] to add array like access
+    T operator[](Eigen::Index) const {
+        return val;
+    }
+    /// size for compatibility
+    Eigen::Index size() const {
+        return 1;
+    }
+private:
+    T val;
+};
+
+/// Meta function to endow floating point array compatible
+template<typename T>
+using makeTypeCompatible = std::conditional<isDual<std::decay_t<T>>, EigenVectorAdaptor<T>, T>;
```

Comment:
You should not need `std::decay` here because `isDual` already does this behind the scenes (removes cv, reference, volatile using `plain<T>` template alias).

Reply:
Yes, that’s true. I’ll fix it at the evening.